### ISSUE TITLE
`Get-Command -CommandType Application` behaves inconsistently, depending on whether a literal or a wildcard-based name is specified

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#24567]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

For background information, see [this Stack Overflow answer](https://stackoverflow.com/a/79175984/45375).

 * For better or, arguably, worse (#12632), `Get-Command foo.txt -Type Application` _does_ find a file by that literal name via `$env:PATH`, whereas `Get-Command foo*.txt -Type Application` does _not_: Due to the latter using a _wildcard_ expression, extensions other than the ones listed in `$env:PATHEXT` are _ignored_.

* Similarly, `Get-Command foo -Type Application` _does_ find extension-less files by that literal name via `$env:PATH`, whereas `Get-Command foo* -Type Application` does _not_: Due to the latter using a _wildcard_ expression, extensions-less files are _ignored_.

### Expected behavior

```console
Consistent treatment of `-Name` arguments, whether they're wildcard-based or not.
```

### Actual behavior

```console
See the description at the top.
```

### Environment data

```powershell
PowerShell 7.5.0-preview.5
```

**Comment by cmkb3** [comment]
This seems tricky to navigate; both interactions make sense.

What 'feels' right to me would be for both wildcard and literal searched to use the current wildcard method -- that is, verifying against PATHEXT -- where indirect ShellExecute/CreateProcess applications (re: anything outside PATHEXT) could be instead be included with an optional `-Indirect` switch or equivalent.

**Comment by mklement0** [comment]
I agree; more generally, such name searches - whether literal or wildcard-based - should only ever match true _executables_ (#12625), which means:

* on Windows: only files that have an extension listed in `$env:PATHEXT`
* on Unix: only files - whatever their extension, including (and typically) _none_  - that are _marked as executable_ (`chmod a+x ...`, in the simplest form), from the caller's perspective.

Complementarily, the ability to actually _execute_ files in `$env:PATH` directories _by name only_ should be limited to true executables (#12632).

Resolving both linked issues as argued there and restated here is what makes the most sense to me personally, as - hopefully - benign bucket-3 breaking changes; neither issue has been formally ruled on.

The purpose of _this_ issue was to point out that the current behavior is _internally inconsistent_ and that even if the current behavior of treating _documents_ as if they were executables is retained, a fix is needed.

---

If the resolution I favor gets implemented, I wouldn't be opposed to providing an _opt-in_ mechanism (such as your `-Indirect` proposal) to allow locating _documents_ (non-executables) in `$env:PATH` directories too, but I wonder how useful that would be.

**Comment by MatejKafka** [comment]
One usecase I can see for the behavior is (mis)using `Get-Command` to lookup .dll dynamic libraries in PATH. Not that I think it's a good idea, but there's probably *someone* using it this way.

**Comment by mklement0** [comment]
Fair point, @MatejKafka.

A pragmatic solution to that would be to special-case `.dll` files and include them in the lookup _if specified by full, literal name_ - which is the only way you can currently discover them (which is one aspect of the inconsistency, but it could be kept for backward compatibility).

This strikes me as an acceptable inconsistency, given that it's unlikely that user will want to _invoke_ a `.dll` file directly (which is currently technically possible, but results in a `No application is associated with the specified file for this operation` error).
